src/libostree/ostree-mutable-tree.c

```diff
@@ -299,6 +446,74 @@ ostree_mutable_tree_ensure_parent_dirs (OstreeMutableTree  *self,
   return ret;
 }
 
+const char empty_tree_csum[] = "6e340b9cffb37a989ca544e6bb780a2c78901d3fb33738768511a30617afa01d";
+
+static void
+set_string(gchar **out, const gchar *in)
+{
+  if (g_strcmp0(*out, in) == 0)
+    return;
+  g_free (*out);
+  *out = g_strdup (in);
+}
+
+/**
+ * ostree_mutable_tree_fill_empty_from_dirtree:
+ *
+ * Merges @self with the tree given by @contents_checksum and
+ * @metadata_checksum, but only if it's possible without writing new objects to
+ * the @repo.  We can do this if either @self is empty, the tree given by
+ * @contents_checksum is empty or if both trees already have the same
+ * @contents_checksum.
+ *
+ * Returns: @TRUE if merge was successful, @FALSE if it was not possible.
+ *
+ * This function enables optimisations when composing trees.  The provided
+ * checksums are not loaded or checked when this function is called.  Instead
+ * the contents will be loaded only when needed.
+ */
+gboolean
+ostree_mutable_tree_fill_empty_from_dirtree (OstreeMutableTree *self,
+                                             OstreeRepo        *repo,
+                                             const char        *contents_checksum,
+                                             const char        *metadata_checksum)
+{
+  g_return_val_if_fail (repo, FALSE);
+
+  switch (self->state)
+    {
+    case MTREE_STATE_LAZY:
+      {
+        if (g_strcmp0 (contents_checksum, self->contents_checksum) == 0 ||
+            g_strcmp0 (empty_tree_csum, self->contents_checksum) == 0)
+          break;
+
+        if (g_strcmp0 (empty_tree_csum, contents_checksum) == 0)
+          {
+            /* Adding an empty tree to a full one - stick with the old contents */
+            contents_checksum = self->contents_checksum;
+            break;
+          }
+        else
+          return FALSE;
+      }
+    case MTREE_STATE_WHOLE:
+      if (g_hash_table_size (self->files) == 0 &&
+          g_hash_table_size (self->subdirs) == 0)
+        break;
+      /* We're not empty - can't convert to a LAZY tree */
+      return FALSE;
+    default:
+      g_return_val_if_reached (FALSE);
+    }
+
+  self->state = MTREE_STATE_LAZY;
+  g_set_object (&self->repo, repo);
+  set_string (&self->contents_checksum, contents_checksum);
+  set_string (&self->metadata_checksum, metadata_checksum);
```

Comment:
So I need to invalidate the *parent* checksum here.  Or have a LAZY tree that (publically) has a NULL checksum.

Reply:
Yep - this was the issue. I've fixed it now.